From c766729572889c48b8022f8fd39e4806e9ad4f5b Mon Sep 17 00:00:00 2001
From: Hiroki Nakagawa <nhiroki@chromium.org>
Date: Tue, 12 Nov 2024 02:32:05 +0000
Subject: [PATCH] [M131 merge] Prerender: Don't call DidStopLoading on activation when it's not reached

> Background of this change

Load events on a prerendered page are deferred until prerender
activation by design. WebContentsObserver::DidStopLoading is one of the
events.

When a prerendered page is loaded, FrameTree::Delegate::DidStopLoading
is called. In non-prerendering cases, the delegate is WebContentsImpl
that notifies WebContentsObserver::DidStopLoading. On the other hand,
in prerendering cases, the delegate is PrerenderHost that ignores the
event, and instead RenderFrameHostImpl::DidCommitPageActivation
*unconditionally* calls DidStopLoading when a prerendered page gets
activated. Deferral of DidStopLoading is implemented in this way.

This should work well when the prerendered page has already reached
DidStopLoading before activation, but does not work when the page
hasn't reached it yet. DidCommitPageActivation notifies the observers of
DidStopLoading, while the page is still being loaded. This results in
confusing the loading state management.

> Solution

To fix this, this CL invokes DidStopLoading on activation only when a
prerendered page has already reached the milestone. The condition is
tracked with DocumentAssociatedData.

(cherry picked from commit 8b0fc98526c83ad801481ac289323e70ef2d6e10)

Change-Id: I40adff7074eaa51fb74692a112ad10809bed5ff0
Bug: 40256454
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5980111
Reviewed-by: Rakina Zata Amni <rakina@chromium.org>
Commit-Queue: Hiroki Nakagawa <nhiroki@chromium.org>
Reviewed-by: Lingqi Chi <lingqi@chromium.org>
Reviewed-by: Kouhei Ueno <kouhei@chromium.org>
Cr-Original-Commit-Position: refs/heads/main@{#1378130}
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/6013641
Commit-Queue: Kouhei Ueno <kouhei@chromium.org>
Cr-Commit-Position: refs/branch-heads/6778@{#2168}
Cr-Branched-From: b21671ca172dcfd1566d41a770b2808e7fa7cd88-refs/heads/main@{#1368529}
---

--- a/chrome/browser/preloading/prefetch/search_prefetch/search_preload_unified_browsertest.cc
+++ b/chrome/browser/preloading/prefetch/search_prefetch/search_preload_unified_browsertest.cc
@@ -2073,8 +2073,12 @@
   content::test::PrerenderHostObserver prerender_observer(
       *GetActiveWebContents(), expected_prerender_url);
 
-  // 4. Activate the prerendered page.
-  NavigateToPrerenderedResult(expected_prerender_url);
+  // 4. Activate the prerendered page. Don't wait for completion of page load,
+  // as it will not be completed until the remaining response is dispatched on
+  // the step 6.
+  NavigateToPrerenderedResult(
+      expected_prerender_url,
+      content::TestNavigationObserver::WaitEvent::kNavigationFinished);
 
   prerender_observer.WaitForActivation();
 
--- a/content/browser/preloading/prerender/prerender_browsertest.cc
+++ b/content/browser/preloading/prerender/prerender_browsertest.cc
@@ -9248,6 +9248,109 @@
   EXPECT_EQ(web_contents()->GetLastCommittedURL(), kPrerenderingUrl);
 }
 
+// Test that WebContentsObserver::DidStopLoading is not invoked when the page
+// gets loaded while prerendering but is invoked on prerender activation.
+IN_PROC_BROWSER_TEST_F(PrerenderBrowserTest,
+                       DidStopLoadingInvokedOnActivation) {
+  const GURL initial_url = GetUrl("/empty.html");
+  const GURL prerendering_url = GetUrl("/simple_page.html");
+
+  web_contents_impl()->set_minimum_delay_between_loading_updates_for_testing(
+      base::Milliseconds(0));
+
+  // Navigate to an initial page.
+  ASSERT_TRUE(NavigateToURL(shell(), initial_url));
+
+  // Initialize a MockWebContentsObserver and ensure that DidStopLoading is
+  // not invoked while prerendering.
+  testing::NiceMock<MockWebContentsObserver> observer(shell()->web_contents());
+  EXPECT_CALL(observer, DidStopLoading()).Times(0);
+
+  // Start a prerender.
+  FrameTreeNodeId prerender_host_id = AddPrerender(prerendering_url);
+  ASSERT_TRUE(prerender_host_id);
+
+  // Verify and clear all expectations on the mock observer before setting new
+  // ones.
+  testing::Mock::VerifyAndClearExpectations(&observer);
+
+  // Activate the prerendered page. This should result in invoking
+  // DidStopLoading.
+  EXPECT_CALL(observer, DidStopLoading()).Times(1);
+  test::PrerenderHostObserver host_observer(*web_contents(), prerender_host_id);
+  NavigatePrimaryPage(prerendering_url);
+  EXPECT_TRUE(host_observer.was_activated());
+  EXPECT_EQ(web_contents()->GetLastCommittedURL(), prerendering_url);
+}
+
+// Test that WebContentsObserver::DidStopLoading is invoked when the page gets
+// loaded after activation.
+//
+// This is a regression test for https://crbug.com/40256454. Previously,
+// DidStopLoading was invoked regardless of the current loading state on
+// activation, which was obviously wrong. This test makes sure that
+// DidStopLoading is invoked after the page actually gets loaded.
+IN_PROC_BROWSER_TEST_F(PrerenderBrowserTest,
+                       DidStopLoadingInvokedAfterActivation) {
+  // Use ControllableHttpResponse to control the timing to serve for delaying
+  // page loading.
+  net::test_server::ControllableHttpResponse response(embedded_test_server(),
+                                                      "/simple_page.html");
+  ASSERT_TRUE(embedded_test_server()->Start());
+
+  const GURL initial_url = embedded_test_server()->GetURL("/empty.html");
+  const GURL prerendering_url =
+      embedded_test_server()->GetURL("/simple_page.html");
+
+  web_contents_impl()->set_minimum_delay_between_loading_updates_for_testing(
+      base::Milliseconds(0));
+
+  // Navigate to an initial page.
+  ASSERT_TRUE(NavigateToURL(shell(), initial_url));
+
+  // Initialize a MockWebContentsObserver and ensure that DidStopLoading is
+  // not invoked while prerendering.
+  testing::NiceMock<MockWebContentsObserver> observer(shell()->web_contents());
+  EXPECT_CALL(observer, DidStopLoading()).Times(0);
+
+  // Start a prerender.
+  AddPrerenderAsync(prerendering_url);
+
+  // Send only a header. This prevents the prerendered page from being fully
+  // loaded.
+  const char kHttpResponseHeader[] =
+      "HTTP/1.1 200 OK\r\n"
+      "Content-Type: text/html; charset=utf-8\r\n"
+      "\r\n";
+  response.WaitForRequest();
+  response.Send(kHttpResponseHeader);
+
+  // Activate the prerendered page. The response for the prerendered page is not
+  // finalized yet, so this should not invoke DidStopLoading.
+  EXPECT_CALL(observer, DidStopLoading()).Times(0);
+  test::PrerenderHostObserver host_observer(*web_contents(), prerendering_url);
+  prerender_helper()->NavigatePrimaryPageAsync(prerendering_url);
+  host_observer.WaitForActivation();
+  EXPECT_TRUE(host_observer.was_activated());
+  EXPECT_EQ(web_contents()->GetLastCommittedURL(), prerendering_url);
+
+  // Verify and clear all expectations on the mock observer before setting new
+  // ones. The activated page is not loaded yet, so DidStopLoading should not be
+  // invoked yet.
+  testing::Mock::VerifyAndClearExpectations(&observer);
+
+  // Finalize the response for the activated page. This should result in
+  // invoking DidStopLoading.
+  base::RunLoop run_loop;
+  EXPECT_CALL(observer, DidStopLoading()).WillOnce(testing::Invoke([&]() {
+    run_loop.Quit();
+  }));
+  response.Send("0\r\n");
+  response.Send("\r\n");
+  response.Done();
+  run_loop.Run();
+}
+
 // Test the dispatch order of various load events on prerender activation.
 IN_PROC_BROWSER_TEST_F(PrerenderBrowserTest, OrderingOfDifferentLoadEvents) {
   const GURL kInitialUrl = GetUrl("/empty.html");
--- a/content/browser/renderer_host/document_associated_data.h
+++ b/content/browser/renderer_host/document_associated_data.h
@@ -90,6 +90,16 @@
     pending_did_finish_load_url_for_prerendering_.reset();
   }
 
+  // Indicates whether `RenderFrameHostImpl::DidStopLoading` was called for this
+  // document while it's prerendering. This is used to defer and dispatch
+  // `WebContentsObserver::DidStopLoading` notification on prerender activation.
+  bool pending_did_stop_loading_for_prerendering() const {
+    return pending_did_stop_loading_for_prerendering_;
+  }
+  void set_pending_did_stop_loading_for_prerendering() {
+    pending_did_stop_loading_for_prerendering_ = true;
+  }
+
   // Reporting API:
   //
   // Contains the reporting source token for this document, which will be
@@ -164,6 +174,7 @@
   bool dom_content_loaded_ = false;
   bool is_discarded_ = false;
   std::optional<GURL> pending_did_finish_load_url_for_prerendering_;
+  bool pending_did_stop_loading_for_prerendering_ = false;
   std::vector<raw_ptr<internal::DocumentServiceBase, VectorExperimental>>
       services_;
   scoped_refptr<NavigationOrDocumentHandle> navigation_or_document_handle_;
--- a/content/browser/renderer_host/render_frame_host_impl.cc
+++ b/content/browser/renderer_host/render_frame_host_impl.cc
@@ -5611,8 +5611,9 @@
   // If any load events occurred pre-activation and were deferred until
   // activation, dispatch them now. This must happen before DidStopLoading() is
   // called because observers expect them to occur before that.
-  if (is_prerender_page_activation)
+  if (is_prerender_page_activation) {
     GetPage().MaybeDispatchLoadEventsOnPrerenderActivation();
+  }
 
   // Try to dispatch DidStopLoading event (note that
   // RenderFrameHostImpl::DidStopLoading implementation won't dispatch the event
@@ -5620,7 +5621,21 @@
   // dispatched pre-activation because the back-forward cache page is already
   // loaded, whereas, for initial prerendering navigation, prerendered page
   // might still be loading.
-  DidStopLoading();
+  if (is_prerender_page_activation) {
+    // Regardless of the actual loading state of the prerendered page, the state
+    // should be overwritten with LoadingState::LOADING_UI_REQUESTED in
+    // SetNavigationRequest().
+    CHECK_EQ(loading_state_, LoadingState::LOADING_UI_REQUESTED);
+
+    // Dispatch DidStopLoading only when the page is fully loaded during
+    // prerendering. Otherwise, DidStopLoading will be dispatched later.
+    if (document_associated_data_
+            ->pending_did_stop_loading_for_prerendering()) {
+      DidStopLoading();
+    }
+  } else {
+    DidStopLoading();
+  }
 
   if (is_prerender_page_activation) {
     // Record metric to check navigation time with prerender activation.
@@ -8953,6 +8968,14 @@
     return;
   }
 
+  // During prerendering, DidStopLoading is dispatched to
+  // PrerenderHost::DidStopLoading, not WebContentsObserver::DidStopLoading. To
+  // dispatch the event to WebContentsObserver on activation later, set the flag
+  // here.
+  if (lifecycle_state() == LifecycleStateImpl::kPrerendering) {
+    document_associated_data_->set_pending_did_stop_loading_for_prerendering();
+  }
+
   was_discarded_ = false;
   loading_state_ = LoadingState::NONE;
 
